server/core/region.go

```diff
@@ -670,6 +686,42 @@ func (r *RegionsInfo) RemoveRegion(region *RegionInfo) {
 	}
 }
 
+type peerSlice []*metapb.Peer
+
+func (s peerSlice) Len() int {
+	return len(s)
+}
+func (s peerSlice) Swap(i, j int) {
+	s[i], s[j] = s[j], s[i]
+}
+func (s peerSlice) Less(i, j int) bool {
+	return s[i].GetId() < s[j].GetId()
+}
+
+// shouldRemoveFromSubTree return true when the region leader changed, peer transferred,
+// new peer was created, learners changed, pendingPeers changed, and so on.
+func (r *RegionsInfo) shouldRemoveFromSubTree(region *RegionInfo, origin *RegionInfo) bool {
+	checkPeersChange := func(origin []*metapb.Peer, other []*metapb.Peer) bool {
+		if len(origin) != len(other) {
+			return true
+		}
+		sort.Sort(peerSlice(origin))
+		sort.Sort(peerSlice(other))
+		for index, peer := range origin {
+			if peer.GetStoreId() == other[index].GetStoreId() && peer.GetId() == other[index].GetId() {
+				continue
+			}
+			return true
+		}
+		return false
+	}
+
+	return origin.leader.GetId() != region.leader.GetId() ||
```

Comment:
When the size or key number of a region changes, does it also need to be removed from subtree?

Reply:
This situation does not need to be removed from the subtree. Because there is processing logic about size and keys number in the (rst *regionSubTree) update method of region.go;

```
func (rst *regionSubTree) update(region *RegionInfo) {
	if r := rst.find(region); r != nil {
		rst.totalSize += region.approximateSize - r.region.approximateSize
		rst.totalKeys += region.approximateKeys - r.region.approximateKeys
		r.region = region
		return
	}
	rst.totalSize += region.approximateSize
	rst.totalKeys += region.approximateKeys
	rst.regionTree.update(region)
}
```

Reply:
ptal @rleungx

Reply:
The size and key number that I mentioned is not a good example. But you can see when we handle the region's heartbeat, there are several cases that we need to update the cache like flow changes and so on. The region will be removed from the subtree in an original way, but it won't now and some function like `GetStoreRegions` may be affected?